### PR TITLE
fix(clickhouse): Reduce attribute metadata cache TTL to fix e2e flakiness

### LINKED ISSUE #8780: ClickHouse e2e tests spurious failures
This happened several times in the past month, the test appears to be unstable

https://github.com/jaegertracing/jaeger/actions/runs/27478516734/job/81222093956?pr=8779

```
=== FAIL: cmd/jaeger/internal/integration TestClickHouseStorage/FindTraces (unknown)
```

Another failure example https://github.com/jaegertracing/jaeger/actions/runs/28978812405/job/85992488652?pr=8976. Here the FindTraces fails for 100 iterations - long enough to address any sort of materialized view delays. More likely this is due to failures during span writing, at least this was the case with similar issues in other storage implementations.

### PR DESCRIPTION
**Pull request by iammdzaidalam**
## Which problem is this PR solving?
- Fixes #8780. The OTel batch processor can flush traces in partial batches, causing `attrMetaCache` to capture an incomplete snapshot of `attribute_metadata` and hold it for up to 1 hour. Queries during this window miss traces at attribute levels not yet present in the cache.

## Description of the changes
- Set `attribute_metadata_cache_ttl: 5s` in `config-clickhouse.yaml` so the integration test's retry loop naturally outlasts any stale cache window
- Reduce `defaultAttributeMetadataCacheTTL` from `time.Hour` to `10 * time.Minute` in `config.go` to shrink the blast radius in production
- Keep `TestBuildStringAttributeCondition_PartialMetadata` to document the symptom, asserting the expected behavior when partial metadata is cached

## How was this change tested?
- TestBuildStringAttributeCondition_PartialMetadata pins the behavior of buildStringAttributeCondition under partial metadata
- Ran STORAGE=clickhouse make jaeger-v2-storage-integration-test locally, all FindTraces subtests pass

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully: `make lint test`

## AI Usage in this PR
- [x] **Moderate**: AI helped with code generation or debugging specific parts

### CODE REVIEW

**Review by yurishkuro — CHANGES_REQUESTED** [review]
I do not understand the change and it looks like AI slop that is gratuitously removing comments.

if we already have the fallback to strings, what exactly is this fix changing?

**Comment by iammdzaidalam** [comment]
Sorry for removing the comments. Total oversight on my part. I just pushed a commit to restore them.

To clarify the fix: you're right there were already two fallbacks, but they only trigger if the tag is missing from the cache entirely, or if parsing fails.

I found a race condition in CI where neither happens. When the batch processor flushes a partial batch, the cache gets a partial entry (e.g. `{my_tag: str/span}`). The original code successfully parses it for the `span` level and skips the final fallback. If the test was actually looking for `my_tag` at the `resource` level, the query only checks `span` attributes and times out returning `0` traces.

Since that partial cache entry lives for an hour, the fix unconditionally applies the full string fallback across all 5 levels every time to avoid this. I also skipped `ValueTypeStr` in the loop since the fallback handles it now.

Kindly take another look when you have a moment.

**Comment by yurishkuro** [comment]
> If the test was actually looking for my_tag at the resource level, the query only checks span attributes and times out returning 0 traces.

do we even have tests that advanced that look at tags at different scopes?

**Comment by iammdzaidalam** [comment]
> > If the test was actually looking for my_tag at the resource level, the query only checks span attributes and times out returning 0 traces.
> 
> do we even have tests that advanced that look at tags at different scopes?

IIRC we have those in fixtures/queries.json

**Comment by iammdzaidalam** [comment]
Yes, `queries.json` has three "Tags in one spot" test cases that check tag lookup at different scopes:

* **Tags:** `span_tags_trace` (tag at the `span` level)
* **Logs:** `log_tags_trace` (tag at the `event` level)
* **Process:** `process_tags_trace` (tag at the `resource` level)

All three use the same `sameplacetag1` key. The first two pass because the stale cache contains entries for both span and event levels, which match those lookups. The third fails because when the cache was first populated, the resource level was not yet present in attribute_metadata... likely because `process_tags_trace` hadnt been flushed yet

**Comment by yurishkuro** [comment]
Thanks for the detailed write-up and the regression test — the test correctly pins the symptom.

However, I think this fix addresses the wrong layer of the problem, and it comes with a permanent query-widening cost that isn't necessary.

## Root cause

The actual bug is in `attribute_metadata.go` at the cache-population site (lines 113–117). `getAttributeMetadata` queries the `attribute_metadata` table and caches the result with a 1-hour TTL. The `attribute_metadata` table is populated **synchronously** by a materialized view on each ClickHouse INSERT — but the OTel batch processor flushes spans in multiple batches. If a query races a flush mid-flight, the cache gets a point-in-time snapshot of the metadata table that reflects only the batches seen so far, and holds it for an hour. Subsequent batches add new `(attribute_key, type, level)` rows to the table, but the cache has no invalidation mechanism, so it never sees them.

## Why the current fix is the wrong direction

Always OR-ing in the 5-level str fallback is a permanent, unconditional cost applied to every attribute filter in every query — even in the steady-state prod case where the cache is correct. In exchange it buys immunity to a transient cache-staleness window. ClickHouse handles wide queries well, but this is still unnecessary work proportional to the number of attribute filters on every FindTraces call, forever.

## Correct fix

The `attribute_metadata_cache_ttl` is already a first-class config field. The fix is:

1. **In `config-clickhouse.yaml`** (used by the integration test), set `attribute_metadata_cache_ttl: 5s`. The existing retry loop in `RunSpanStoreTests` will naturally ride out the TTL: even if a query races a partial flush and caches incomplete metadata, the cache expires in 5s and the next retry picks up the fully-populated table.

2. **In `config.go`**, reduce `defaultAttributeMetadataCacheTTL` from `time.Hour` to `10 * time.Minute`. This doesn't eliminate the prod window, but it shrinks it from 60 minutes to 10 minutes for the edge case of a new attribute key being introduced.

3. **Revert the changes to `query_builder.go`** — the query widening is not needed once the cache heals within the retry window.

The regression test `TestBuildStringAttributeCondition_PartialMetadata` is worth keeping, but it should assert the *old* behavior (partial metadata → only span str level queried), with a comment explaining that the correct remedy is TTL expiry + retry, not query widening.

**Comment by iammdzaidalam** [comment]
Thanks for the detailed explanation. I will update the PR to set a short TTL in the integration test config and reduce the default in config.go, and revert the query builder changes.

File: internal/storage/v2/clickhouse/tracestore/query_builder_test.go [hunk 1]
```diff
@@ -190,3 +190,24 @@ func TestBuildStringAttributeCondition_MultipleTypes(t *testing.T) {
 	assert.Contains(t, query, "str_attributes")
 	assert.Len(t, args, 4)
 }
+
+// Regression test for https://github.com/jaegertracing/jaeger/issues/8780:
+// Note: We intentionally assert the "old" behavior here (only span str level queried).
+// The correct remedy for a partial cache is TTL expiry + retry, not query widening,
+// as widening the query adds a permanent performance cost to all string queries.
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply internal/storage/v2/clickhouse/tracestore/query_builder_test.go]: This is intentional per reviewer feedback. The fix is at the TTL layer (5s in the integration test config), not the query layer. The test documents the symptom so future readers understand why the TTL approach is correct.

File: internal/storage/v2/clickhouse/config.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ const (
 	defaultDatabase                      = "jaeger"
 	defaultSearchDepth                   = 1000
 	defaultMaxSearchDepth                = 10000
-	defaultAttributeMetadataCacheTTL     = time.Hour
+	defaultAttributeMetadataCacheTTL     = 10 * time.Minute
 	defaultAttributeMetadataCacheMaxSize = 1000
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply internal/storage/v2/clickhouse/config.go:22]: PR title and description have been updated to reflect the TTL-based approach. The operational impact was already considered by Yuri who suggested this change.

File: internal/storage/v2/clickhouse/tracestore/query_builder_test.go [hunk 3]
```diff
@@ -190,3 +190,24 @@ func TestBuildStringAttributeCondition_MultipleTypes(t *testing.T) {
...
+	args := buildStringAttributeCondition(&q, nil, "k", attr, partialMetadata)
+
+	query := q.String()
+	assert.Contains(t, query, "str_attributes")
+	assert.NotContains(t, query, "resource_str_attributes")
+	assert.NotContains(t, query, "scope_str_attributes")
+	assert.NotContains(t, query, "events")
+	assert.NotContains(t, query, "links")
+	assert.Len(t, args, 2)
+}
```
[the remark this answers is not in the document]
Comment (iammdzaidalam) [reply internal/storage/v2/clickhouse/tracestore/query_builder_test.go]: Intentional per Yuri's direction, the fix is in the cache TTL, not the query. This test pins the existing query behavior so future readers understand why the TTL approach is preferred over query widening.

**Comment by iammdzaidalam** [comment]
I have pushed the updates and changed the PR title and description.

As requested:

1. Reverted the `query_builder.go` query widening logic and the corresponding SQL snapshots.
2. Reduced `defaultAttributeMetadataCacheTTL` to `10m` in `config.go`.
3. Set `attribute_metadata_cache_ttl: 5s` in `config-clickhouse.yaml`.
4. Kept the regression test but updated it to assert the old behavior, with a comment explaining that TTL expiry and retry is the correct remedy.

Thanks again for the guidance, PTAL when free.

File: cmd/jaeger/config-clickhouse.yaml [hunk 4]
```diff
@@ -39,6 +39,7 @@ extensions:
     backends:
       some-storage:
         clickhouse:
+          attribute_metadata_cache_ttl: 5s
```
Comment (yurishkuro) [inline cmd/jaeger/config-clickhouse.yaml]: this is not purely an integration test config, 5s is unacceptable for prod. We can use env var mechanism to inject a shorter ttl from e2e tests, leaving the default where it was.

**Review by yurishkuro — CHANGES_REQUESTED** [review]

File: cmd/jaeger/config-clickhouse.yaml (hunk 4 above)
Comment (iammdzaidalam) [reply cmd/jaeger/config-clickhouse.yaml]: Switched to the env var mechanism, same as badger's `BADGER_METRICS_UPDATE_INTERVAL`. The yaml now reads `attribute_metadata_cache_ttl: ${env:CLICKHOUSE_ATTR_METADATA_CACHE_TTL:-10m}` and `clickhouse_test.go` propagates the var to the jaeger subprocess. The 5s is exported from `scripts/e2e/clickhouse.sh` so it only applies to the e2e run, and a local run reproduces CI.

BTW i left `defaultAttributeMetadataCacheTTL` at 10m per your earlier comment. Should i restore the 1hr there too?

File: internal/storage/v2/clickhouse/tracestore/query_builder_test.go [hunk 5]
```diff
@@ -190,3 +190,24 @@ func TestBuildStringAttributeCondition_MultipleTypes(t *testing.T) {
...
+	attr := pcommon.NewValueStr("v")
+	var q strings.Builder
+	args := buildStringAttributeCondition(&q, nil, "k", attr, partialMetadata)
+
+	query := q.String()
+	assert.Contains(t, query, "str_attributes")
+	assert.NotContains(t, query, "resource_str_attributes")
+	assert.NotContains(t, query, "scope_str_attributes")
+	assert.NotContains(t, query, "events")
+	assert.NotContains(t, query, "links")
```
Comment (yurishkuro) [inline internal/storage/v2/clickhouse/tracestore/query_builder_test.go]: I don't know what we're testing here. The tests against the query shape must use snapshot testing where the whole query is materialized into a fixture file.

Comment (iammdzaidalam) [reply internal/storage/v2/clickhouse/tracestore/query_builder_test.go]: You're right, the snapshots already cover it. In `TestFindTraceIDs_2.sql`, `no.metadata` has no metadata row and expands to all five levels, while `event.attr` only has an event-level row and produces a single events-only clause (that narrowing is the #8780 symptom). Dropping the test rather than add a redundant fixture.

**Comment by yurishkuro** [comment]
this may have been fixed by the recent split of e2e tests into write and read phase. Reducing cache TTL would actually make it worse.
